core/DataAccess/Model.php

```diff
@@ -354,50 +354,47 @@ public function getArchiveIdsForDeletedSites($archiveTableName)
     {
-        $validSegmentClauses = [];
-
-        foreach ($segmentHashesById as $idSite => $segments) {
-            // segments are md5 hashes and such not a problem re sql injection. for performance etc we don't want to use
-            // bound parameters for the query
-            foreach ($segments as $segment) {
-                if (!preg_match('/^[a-z0-9A-Z]+$/', $segment)) {
-                    throw new Exception($segment . ' expected to be an md5 hash');
-                }
-            }
-
-            // Special case as idsite=0 means the segments are not site-specific
-            if ($idSite === 0) {
-                foreach ($segments as $segmentHash) {
-                    $validSegmentClauses[] = '(name LIKE "done' . $segmentHash . '%")';
-                }
-                continue;
-            }
-
-            $idSite = (int)$idSite;
-
-            // Vanilla case - segments that are valid for a single site only
-            $sql = '(idsite = ' . $idSite . ' AND (';
-            $sql .= 'name LIKE "done' . implode('%" OR name LIKE "done', $segments) . '%"';
-            $sql .= '))';
-            $validSegmentClauses[] = $sql;
+        $segmentClauses = [];
+        foreach ($segments as $segment) {
+            $segmentClauses[] = self::getDeletedSegmentWhereClause($segment);
         }
 
-        $isValidSegmentSql = implode(' OR ', $validSegmentClauses);
+        $segmentClauses = implode(' OR ', $segmentClauses);
 
         $sql = 'SELECT idarchive FROM ' . $archiveTableName
-            . ' WHERE name LIKE "done%" AND name != "done"'
-            . ' AND ts_archived < ?'
-            . ' AND NOT (' . $isValidSegmentSql . ')';
+            . ' WHERE ts_archived < ?'
+            . ' AND (' . $segmentClauses . ')';
 
         $rows = Db::fetchAll($sql, array($oldestToKeep));
 
-        return array_map(function($row) { return $row['idarchive']; }, $rows);
+        return array_column($rows, 'idarchive');
+    }
+
+    private static function getDeletedSegmentWhereClause(array $segment)
+    {
+        $idSite = (int)$segment['enable_only_idsite'];
+        $segmentHash = Segment::getSegmentHash($segment['definition']);
+        // Valid segment hashes are md5 strings - just confirm that it is so it's safe for SQL injection
+        if (!preg_match('/^[a-z0-9A-Z]+$/', $segmentHash)) {
+            throw new Exception($segment . ' expected to be an md5 hash');
+        }
+
+        $nameClause = 'name LIKE "done' . $segmentHash . '%"';
+        $idSiteClause = '';
+        if ($idSite > 0) {
+            $idSiteClause = ' AND idsite = ' . $idSite;
+        } elseif (! empty($segment['idsites_to_preserve'])) {
+            // A segment for all sites was deleted, but there are segments for a single site with the same definition
+            $idSiteClause = ' AND idsite NOT IN (' . implode(',', $segment['idsites_to_preserve']) . ')';
```

Comment:
To be safe re possible SQL injection we should do `$segment['idsites_to_preserve'] = array_map('intval', $segment['idsites_to_preserve'])`